collector/wifi_linux.go

```diff
@@ -0,0 +1,281 @@
+// limitations under the License.
+
+package collector
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/mdlayher/wifi"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type wifiCollector struct {
+	InterfaceFrequencyHertz *prometheus.Desc
+
+	StationConnectedSecondsTotal *prometheus.Desc
+	StationInactiveSeconds       *prometheus.Desc
+	StationReceiveBitrateBPS     *prometheus.Desc
+	StationTransmitBitrateBPS    *prometheus.Desc
+	StationSignalDBM             *prometheus.Desc
+	StationTransmitRetriesTotal  *prometheus.Desc
+	StationTransmitFailedTotal   *prometheus.Desc
+	StationBeaconLossTotal       *prometheus.Desc
+
+	stat wifiStater
+}
+
+var (
+	collectorWifi = flag.String("collector.wifi", "", "test fixtures to use for wifi collector metrics")
+)
+
+func init() {
+	Factories["wifi"] = NewWifiCollector
+}
+
+var _ wifiStater = &wifi.Client{}
+
+// wifiStater is an interface used to swap out a *wifi.Client for end to end tests.
+type wifiStater interface {
+	Interfaces() ([]*wifi.Interface, error)
+	StationInfo(ifi *wifi.Interface) (*wifi.StationInfo, error)
+}
+
+func NewWifiCollector() (Collector, error) {
+	stat, err := newWifiStater(*collectorWifi)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access wifi data: %v", err)
+	}
+
+	const (
+		subsystem = "wifi"
+	)
+
+	var (
+		labels = []string{"device"}
+	)
+
+	return &wifiCollector{
+		InterfaceFrequencyHertz: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "interface_frequency_hertz"),
+			"The current frequency a WiFi interface is operating at, in hertz.",
+			labels,
+			nil,
+		),
+
+		StationConnectedSecondsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "station_connected_seconds_total"),
+			"The total number of seconds a station has been connected to an access point.",
+			labels,
+			nil,
+		),
+
+		StationInactiveSeconds: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "station_inactive_seconds"),
+			"The number of seconds since any wireless activity has occurred on a station.",
+			labels,
+			nil,
+		),
+
+		StationReceiveBitrateBPS: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "station_receive_bitrate_bps"),
+			"The current WiFi receive bitrate of a station, in bits per second.",
+			labels,
+			nil,
+		),
+
+		StationTransmitBitrateBPS: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "station_transmit_bitrate_bps"),
+			"The current WiFi transmit bitrate of a station, in bits per second.",
+			labels,
+			nil,
+		),
+
+		StationSignalDBM: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "station_signal_dbm"),
+			"The current WiFi signal strength, in decibel-milliwatts (dBm).",
+			labels,
+			nil,
+		),
+
+		StationTransmitRetriesTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "station_transmit_retries_total"),
```

Comment:
Unfortunately nl80211 isn't super clear about it: http://lxr.free-electrons.com/source/include/uapi/linux/nl80211.h#L2620.

```
* @NL80211_STA_INFO_TX_FAILED: total failed packets (MPDUs)
*      (u32, to this station)
```

For the time being, at least it'll be useful to see if/when a number of transmission failures and beacon losses start occurring, so a weak connection can be easily identified.